### ISSUE TITLE
Segmentation fault error on Mac

### ISSUE THREAD

**Issue by carobs9** [MaartenGr/BERTopic#2359]
### Have you searched existing issues?  🔎

- [x] I have searched and found no existing issues

### Desribe the bug

Hello, 

I am trying to train a simple BERTopic model using available GPU, but I keep running into a segmentation fault and a leaked semaphore warning, to the point that the code does not compile. I am fitting (.fit) a model on 5,000 representative short descriptions and calling .transform on 368,049 short descriptions. I believe the lack of RAM should not be a problem, since I have tested the code with a small subset of the data, and I still run into the same issue.

### Reproduction

I am using Python 3.12.10, and training on a 2019 MacPro with the following specifications: 
- Processor: 2,7 GHz 24-Core Intel Xeon W
- Memory: 192 GB 2933 MHz DDR4
- macOS 15.4

Here is the code I use:
```python
from bertopic import BERTopic
import numpy as np
import config as cfg
from data_loader import DataLoader
import utils
from topic_modeler import TopicModeler
from umap import UMAP
from visualizer import Visualizer
from sentence_transformers import SentenceTransformer

'''Make sure SUBSETTING is set to False in config'''

device = utils.get_device()  # get device
print(f"Running on {device}") 

data_loader = DataLoader(cfg)
topic_modeler = TopicModeler(cfg)
df, docs = data_loader.load_data()

subset = docs[0:5000]
print(df.shape)
print(len(subset))

topic_model = BERTopic(
                embedding_model=topic_modeler.embedding_model, 
                vectorizer_model=topic_modeler.vectorizer_model,
                language=topic_modeler.language,
                calculate_probabilities=True,
                verbose=True,
                umap_model=UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', low_memory=True),
                min_topic_size=topic_modeler.min_topic_size,
                hdbscan_model=topic_modeler.hdbscan_model if cfg.CUSTOM_HDBSCAN else None
            )

topic_model.fit(subset)
embeddings = SentenceTransformer(cfg.EMBEDDING_MODEL).encode(docs)
np.save(cfg.MODELS_PATH / f'TEST_embeddings_{cfg.MODEL_NAME}.npy', embeddings)
topics, probs = topic_model.transform(docs, embeddings=embeddings)
```

This is the error I get:
```python
BERTopic_training/finetuning.py
Running on mps
No sentence-transformers model found with name sentence-transformers/all-MiniLM-L6-v2. Creating a new one with mean pooling.
Loading unlabelled dataset: all_missions → Shape: (368049, 11)
TRUNCATION DISABLED → Using original dataset format.
FINAL DATA SHAPE: (368049, 11)
(368049, 11)
5000
2025-05-20 12:44:35,637 - BERTopic - Embedding - Transforming documents to embeddings.
Batches: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 157/157 [00:17<00:00,  9.21it/s]
2025-05-20 12:44:52,799 - BERTopic - Embedding - Completed ✓
2025-05-20 12:44:52,800 - BERTopic - Dimensionality - Fitting the dimensionality reduction algorithm
OMP: Info #271: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
zsh: segmentation fault (...)/lastenv/bin/python 
/usr/local/Cellar/python@3.12/3.12.10/Frameworks/Python.framework/Versions/3.12/lib/python3.12/multiprocessing/resource_tracker.py:279: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown                                         
  warnings.warn('resource_tracker: There appear to be %d
```

Here are the dependencies used:

Package      ,         Version
--------------------- -----------
adjustText            1.3.0
annotated-types       0.7.0
anyio                 4.9.0
appdirs               1.4.4
bertopic              0.17.0
blis                  1.2.1
bokeh                 3.7.3
catalogue             2.0.10
certifi               2025.4.26
charset-normalizer    3.4.2
click                 8.1.8
cloudpathlib          0.21.1
cloudpickle           3.1.1
colorcet              3.1.0
colorspacious         1.1.2
confection            0.1.5
contourpy             1.3.2
cycler                0.12.1
cymem                 2.0.11
dask                  2024.12.1
dask-expr             1.1.21
datamapplot           0.5.1
datashader            0.18.1
distributed           2024.12.1
docker-pycreds        0.4.0
fastapi               0.110.0
filelock              3.18.0
fonttools             4.58.0
fsspec                2025.3.2
geopandas             1.0.1
gitdb                 4.0.12
GitPython             3.1.44
hdbscan               0.8.40
huggingface-hub       0.31.4
idna                  3.10
imageio               2.37.0
importlib_resources   6.5.2
Jinja2                3.1.6
joblib                1.5.0
kiwisolver            1.4.8
langcodes             3.5.0
language_data         1.3.0
lazy_loader           0.4
llvmlite              0.44.0
locket                1.0.0
lz4                   4.4.4
marisa-trie           1.2.1
markdown-it-py        3.0.0
MarkupSafe            3.0.2
matplotlib            3.10.3
mdurl                 0.1.2
mpmath                1.3.0
msgpack               1.1.0
multipledispatch      1.0.0
murmurhash            1.0.12
narwhals              1.40.0
networkx              3.4.2
nltk                  3.9.1
numba                 0.61.2
numpy                 1.26.4
packaging             25.0
pandas                2.2.3
param                 2.2.0
partd                 1.4.2
pillow                11.2.1
pip                   25.0.1
platformdirs          4.3.8
plotly                6.1.0
preshed               3.0.9
protobuf              6.31.0
psutil                7.0.0
pyarrow               20.0.0
pyct                  0.5.0
pydantic              2.11.4
pydantic_core         2.33.2
Pygments              2.19.1
pygris                0.2.0
pylabeladjust         0.1.13
pynndescent           0.5.13
pyogrio               0.11.0
pyparsing             3.2.3
pyproj                3.7.1
Pyqtree               1.0.0
python-dateutil       2.9.0.post0
pytz                  2025.2
PyYAML                6.0.2
rcssmin               1.2.1
regex                 2024.11.6
requests              2.32.3
rich                  14.0.0
rjsmin                1.2.4
safetensors           0.5.3
scikit-image          0.25.2
scikit-learn          1.4.2
scipy                 1.15.3
sentence-transformers 4.1.0
sentry-sdk            2.28.0
setproctitle          1.3.6
setuptools            80.7.1
shapely               2.1.1
shellingham           1.5.4
six                   1.17.0
smart-open            7.1.0
smmap                 5.0.2
sniffio               1.3.1
sortedcontainers      2.4.0
spacy                 3.8.6
spacy-legacy          3.0.12
spacy-loggers         1.0.5
srsly                 2.5.1
starlette             0.36.3
sympy                 1.14.0
tblib                 3.1.0
thinc                 8.3.4
threadpoolctl         3.6.0
tifffile              2025.5.10
tokenizers            0.21.1
toolz                 1.0.0
torch                 2.2.2
tornado               6.5
tqdm                  4.67.1
transformers          4.51.3
typer                 0.15.4
typing_extensions     4.13.2
typing-inspection     0.4.0
tzdata                2025.2
umap-learn            0.5.7
urllib3               2.4.0
wandb                 0.19.11
wasabi                1.1.3
weasel                0.4.1
wrapt                 1.17.2
xarray                2025.4.0
xyzservices           2025.4.0
zict                  3.0.0

### BERTopic Version

0.17.0

**Comment by MaartenGr** [comment]
Could you share what exactly is in all those `topic_modeler.` parameters? It's a bit tricky to understand what is happening here considering I can't see any of the models.

**Comment by carobs9** [comment]
Hi Maarten, 

Thank you for your quick answer. Yes, this is the TopicModeler class:

'''python
class TopicModeler:
    """
    Class that wraps the full BERTopic training, saving, and loading workflow.

    Handles configuration-based embedding models, vectorization, dimensionality reduction (UMAP), 
    clustering (HDBSCAN), training, saving, logging (via Weights & Biases), and transformation.
    """
    def __init__(self, cfg):
        """
        Initialize the TopicModeler with a configuration object.

        Parameters:
        ----------
        cfg : object
            Configuration object containing settings such as model name, paths,
            embedding model, flags for custom UMAP/HDBSCAN, etc.
        """
        self.cfg = cfg
        self.embedding_model = SentenceTransformer(self.cfg.EMBEDDING_MODEL)
        self.tokenizer = LemmaTokenizer()
        self.vectorizer_model = CountVectorizer(stop_words="english", tokenizer=self.tokenizer) 
        self.language = "english"
        self.min_topic_size = self.cfg.MIN_TOPIC_SIZE
        self.device = utils.get_device()
        utils.set_seed(self.cfg.SEED) 

        # custom hdbscan settings
        self.hdbscan_model = None
        if self.cfg.CUSTOM_HDBSCAN:
            self.hdbscan_model = HDBSCAN(  
                min_cluster_size=10,
                min_samples=10, # provide a measure of how conservative you want your clustering to be - larger = more conservative
                cluster_selection_epsilon=0.01,
                metric="euclidean",
                cluster_selection_method="eom",
                prediction_data=True
            )

        # custom umap settings
        self.umap_model = None
        if self.cfg.CUSTOM_UMAP:
            self.umap_model = umap.UMAP(
                n_neighbors=15,
                n_components=5,
                min_dist=0.0,
                metric='cosine',
                low_memory=False,
                random_state=self.cfg.SEED() 
            )

        self.topic_model = None

    def get_pre_embeddings(self, docs):
        embeddings = self.embedding_model.encode(docs)
        return embeddings
    
    def save_pre_embeddings(self, embeddings):
        np.save(self.cfg.MODELS_PATH / f'embeddings_{self.cfg.MODEL_NAME}.npy', embeddings)

    def open_pre_embeddings(self):
        path = self.cfg.MODELS_PATH / f"embeddings_{self.cfg.MODEL_NAME}.npy"
        return np.load(path)
    
    def train(self, docs):
        """
        Trains a BERTopic model using the configured settings and input documents.

        Parameters:
        ----------
        docs : list of str
            A list of text documents to train the topic model on.

        Returns:
        -------
        topics : list of int
            Assigned topic for each document.

        probs : list of list of float
            Probability distribution of topics for each document.
        """
        print('Training model with the following parameters:\n',
              f'Embedding model: {self.embedding_model}\n',
              f'Vectorizer: {self.vectorizer_model}\n', 
              f'Language: {self.language}\n',
              f'Minimum topic size: {self.min_topic_size}\n',
              f'Using Custom HDBSCAN: {self.cfg.CUSTOM_HDBSCAN}\n',
              f'Using Custom UMAP: {self.cfg.CUSTOM_UMAP}\n')
        
        self.topic_model = BERTopic(
                embedding_model=self.embedding_model, 
                vectorizer_model=self.vectorizer_model,
                language=self.language,
                calculate_probabilities=True,
                verbose=True,
                min_topic_size=self.min_topic_size,
                hdbscan_model=self.hdbscan_model if self.cfg.CUSTOM_HDBSCAN else None
            )

        if self.hdbscan_model:
            print("Custom HDBSCAN model is being used:", self.hdbscan_model.get_params())
        else:
            print("Using BERTopic's default HDBSCAN settings.")
        if self.umap_model:
            print("Custom UMAP model is being used:", self.umap_model.get_params())
        else:
            print("Using BERTopic's default UMAP settings.")

        self.topic_model.language = "english"

        topics, probs = self.topic_model.fit_transform(docs)

        # setting custom topic labels 
        self.set_topic_labels()

        if self.cfg.USE_WANDB:
            wandb.log({
            "online_training": self.cfg.ONLINE_TRAINING,
            "min_topic_size": self.min_topic_size,
            "min_cluster_size": self.min_cluster_size if self.cfg.CUSTOM_HDBSCAN else 'Default',
            "min_samples": self.min_samples if self.cfg.CUSTOM_HDBSCAN else 'Default',
            "vectorizer": str(self.vectorizer_model)
            })
            
        return topics, probs
'''

And these are the parameters used for the specific training I had issues with:

  Training model with the following parameters:
   - Embedding model: SentenceTransformer(
    (0): Transformer({'max_seq_length': 256, 'do_lower_case': False}) with Transformer model: BertModel 
    (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
    (2): Normalize()
  )
   - Vectorizer: CountVectorizer(stop_words='english',
                  tokenizer=<topic_modeler.LemmaTokenizer object at 0x00000280190C03D0>)
   - Language: english
   - Minimum topic size: 26
   - Using Custom HDBSCAN: False
   - Using Custom UMAP: False

**Comment by carobs9** [comment]
Thank you for your answer @MaartenGr. Using PCA instead of UMAP removed the error, so I believe the UMAP initialization was the problem. After trying with many different package versions unsuccessfully, I ended up running a Linux-based Docker container to solve the issue with the following requirements. Unfortunately, this is just a workaround and is definitely not the best long-term solution because it does not allow to use the available Mac GPU.

```python

annotated-types==0.7.0
anyio==4.9.0
bertopic==0.17.0
bokeh==3.7.3
certifi==2025.4.26
charset-normalizer==3.4.2
click==8.2.1
cloudpickle==3.1.1
colorama==0.4.6
colorcet==3.1.0
colorspacious==1.1.2
contourpy==1.3.2
cycler==0.12.1
dask==2024.12.1
dask-expr==1.1.21
datamapplot==0.5.1
datashader==0.18.1
distributed==2024.12.1
docker-pycreds==0.4.0
fastapi==0.110.0
filelock==3.18.0
fonttools==4.58.0
fsspec==2025.5.0
gitdb==4.0.12
GitPython==3.1.44
hdbscan==0.8.40
huggingface-hub==0.31.4
nbformat==5.10.4
idna==3.10
imageio==2.37.0
importlib_metadata==8.7.0
importlib_resources==6.5.2
Jinja2==3.1.6
joblib==1.5.0
kiwisolver==1.4.8
lazy_loader==0.4
llvmlite==0.43.0
locket==1.0.0
lz4==4.4.4
MarkupSafe==3.0.2
matplotlib==3.10.3
mpmath==1.3.0
msgpack==1.1.0
multipledispatch==1.0.0
narwhals==1.40.0
networkx==3.4.2
nltk==3.9.1
numba==0.60.0
numpy==1.26.4
packaging==25.0
pandas==2.2.3
param==2.2.0
partd==1.4.2
pillow==11.2.1
pip==24.0
platformdirs==4.3.8
plotly==6.1.1
protobuf==6.31.0
psutil==7.0.0
pyarrow==20.0.0
pyct==0.5.0
pydantic==2.11.4
pydantic_core==2.33.2
pylabeladjust==0.1.13
pynndescent==0.5.13
pyparsing==3.2.3
Pyqtree==1.0.0
python-dateutil==2.9.0.post0
pytz==2025.2
PyYAML==6.0.2
rcssmin==1.2.1
regex==2024.11.6
requests==2.32.3
rjsmin==1.2.4
safetensors==0.5.3
scikit-image==0.25.2
scikit-learn==1.3.2
scipy==1.15.3
sentence-transformers==4.1.0
sentry-sdk==2.29.1
setproctitle==1.3.6
setuptools==65.5.0
six==1.17.0
smmap==5.0.2
sniffio==1.3.1
sortedcontainers==2.4.0
starlette==0.36.3
sympy==1.14.0
tblib==3.1.0
threadpoolctl==3.6.0
tifffile==2025.5.26
tokenizers==0.21.1
toolz==1.0.0
torch==2.2.2
tornado==6.5.1
tqdm==4.67.1
transformers==4.51.3
typing_extensions==4.13.2
typing-inspection==0.4.1
tzdata==2025.2
umap-learn==0.5.7
urllib3==2.4.0
wandb==0.19.11
xarray==2025.4.0
xyzservices==2025.4.0
zict==3.0.0
zipp==3.22.0
seaborn 
```

**Comment by Eric-WK** [comment]
Hi, I tried the following approaches:


```
umap_model = umap.UMAP(n_components=2, random_state=42)

umap_model_default = umap.UMAP(
                    n_neighbors=15,
                    n_components=5,
                    min_dist=0.0,
                    metric="cosine",
                    low_memory=True,
                )

pca_model = PCA(n_components=5, random_state=42)


topic_model = BERTopic(
    verbose=True,
    min_topic_size=3,
    umap_model =umap_model_default
    )
topics, probs = topic_model.fit_transform(text, embeddings)
```


the default failed, and the `umap_model` and `pca_model` worked. Looking a bit further, I was able to set the following environment variables: 
```
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["NUMBA_NUM_THREADS"] = "1"
```

which allowed me run it with `umap_model_default` which is the same init as from the `_bertopic.py`.

**Comment by carobs9** [comment]
Hi @Eric-WK ,

Thank you for your answer. Just to make sure I understood correctly: The UMAP with only two components and random state worked initially, and the default UMAP model only worked after setting the environment variables you describe?

**Comment by Eric-WK** [comment]
> Hi [@Eric-WK](https://github.com/Eric-WK) ,
> 
> Thank you for your answer. Just to make sure I understood correctly: The UMAP with only two components and random state worked initially, and the default UMAP model only worked after setting the environment variables you describe?

Hi @carobs9 , yes exactly. I tested it in a jupyter notebook and as a standalone script, both behaved the same way. 

These worked without the environment variables 
```
umap_model = umap.UMAP(n_components=2, random_state=42)
pca_model = PCA(n_components=5, random_state=42)
```

This worked only after setting the environment variables: 
NB: env variables were set at the beginning of the code
```
umap_model_default = umap.UMAP(
                    n_neighbors=15,
                    n_components=5,
                    min_dist=0.0,
                    metric="cosine",
                    low_memory=True,
                )
```

**Comment by carobs9** [comment]
I unfortunately do not have access to the MacPro I used when the initial issue happened, so the exact replication is impossible for me at this moment. On the other hand, I tried to replicate the issue and @Eric-WK 's solution on a 2024 MacBook Air with the following specifications:

- Apple M3 chip
- 16 GB of memory
- macOS Sequoia 15.6.1

The solution proposed worked also in my case. Before setting the environment variables, I obtained the segmentation fault and error and semaphore warning described at the beginning of the thread. After setting the following environment variables, the UMAP model worked normally:

```python
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["NUMBA_NUM_THREADS"] = "1"
